flock_agent/daemon/daemon.py

```diff
@@ -357,3 +361,94 @@ def log(self, request, response, time):
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"installed version: {self.c.version}, latest version: {latest_version}",
+                )
+                if latest_version <= self.c.version:
+                    continue
+
+                # Find the pkg asset
+                url = None
+                filename = None
+                for asset in release["assets"]:
+                    if asset["name"].endswith(".pkg"):
+                        url = asset["browser_download_url"]
+                        filename = asset["name"]
+                        break
+
+                if not url:
+                    self.c.log("Daemon", "autoupdate_loop", "could not find .pkg file")
+                    continue
+
+                # Download the update
+                self.c.log("Daemon", "autoupdate_loop", f"downloading {url}")
+                r = requests.get(url)
+
+                os.makedirs(os.path.join(self.lib_dir, "updates"), exist_ok=True)
+                download_filename = os.path.join(self.lib_dir, "updates", filename)
+                with open(download_filename, "wb") as f:
+                    f.write(r.content)
+
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"download complete: {download_filename}",
+                )
+
+                # Verify that it's codesigned
+                p = subprocess.run(
+                    ["/usr/sbin/pkgutil", "--check-signature", download_filename],
+                    stdout=subprocess.PIPE,
+                )
+                if (
+                    p.returncode != 0
+                    or "Status: signed by a developer certificate issued by Apple for distribution"
+                    not in p.stdout.decode()
+                    or "Developer ID Installer: FIRST LOOK PRODUCTIONS"
+                    not in p.stdout.decode()
+                ):
+                    self.c.log(
+                        "Daemon",
+                        "autoupdate_loop",
+                        f"codesign verification failed: {p.stdout.decode()}",
+                    )
+                    continue
+
+                # Install the update
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    "launching installer background process and quitting daemon",
+                )
+                subprocess.Popen(
+                    ["/usr/sbin/installer", "-pkg", download_filename, "-target", "/"]
+                )
+                sys.exit(0)
+
+            except Exception as e:
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"Exception while checking for updates: {e}",
+                )
+
+            # Wait 12 hours
+            await asyncio.sleep(43200)
```

Comment:
Possible future improvement: faster retries with exponential back off. Also after a successful update, maybe check more frequently, so that if something goes wrong we're able to fix it faster.

Reply:
That's a good idea. For now though the update check basically just loads a URL from github -- we could do this more frequently, like every 6 hours instead. Think it's worth it or this is fine?

Reply:
I think this is probably fine for now.